### ISSUE TITLE
vmagent issues with proxy_url

### ISSUE THREAD

**Issue by ThomasADavis** [VictoriaMetrics/VictoriaMetrics#1179]
**Describe the bug**

I'm trying to get vmagent, to scrape a prometheus push proxy.

https://github.com/prometheus-community/PushProx

With prometheus, it works correctly.

With vmagent, I get 

```
error  when scraping "http://dev-k3s-2.omni:9100/metrics": error when sending  CONNECT request to proxy "http://dev-k3s-2:8080/": unexpected status  code received: 500; want: 200; try -enableTCP6 command-line flag if you  scrape ipv6 addresses
```

**To Reproduce**

Go to https://github.com/prometheus-community/PushProx.  Follow the steps to set up a simple proxy/client.

The proxy/client config there assumes a node-exporter is running.  It can be any exporter.

**Expected behavior**

vmagent able to scrape the proxy the same as prometheus.

**Screenshots**

**Version**
```
[root@dev-k3s-2 v]# ./vmagent-prod --version
vmagent-20210330-124411-tags-v1.57.1-0-g2d3082bb5
```

**Used command-line flags**

```
[root@dev-k3s-2 v]# ./vmagent-prod -remoteWrite.tmpDataPath /var/lib/metrics/ -promscrape.config /etc/prometheus/prometheus.yml -remoteWrite.url http://omni-vmetric.omni/insert/100000/prometheus/api/v1/write
```

Prometheus config:

```
[root@dev-k3s-2 v]# cat /etc/prometheus/prometheus.yml 
# my global config
global:
  scrape_interval:     15s # Set the scrape interval to every 15 seconds. Default is every 1 minute.
  evaluation_interval: 15s # Evaluate rules every 15 seconds. The default is every 1 minute.
  # scrape_timeout is set to the global default (10s).

# Alertmanager configuration
alerting:
  alertmanagers:
  - static_configs:
    - targets:
      # - alertmanager:9093

# Load rules once and periodically evaluate them according to the global 'evaluation_interval'.
rule_files:
  # - "first_rules.yml"
  # - "second_rules.yml"

# A scrape configuration containing exactly one endpoint to scrape:
# Here it's Prometheus itself.
scrape_configs:
  # The job name is added as a label `job=<job_name>` to any timeseries scraped from this config.
  - job_name: 'prometheus'

    # metrics_path defaults to '/metrics'
    # scheme defaults to 'http'.

    static_configs:
    - targets: ['localhost:9090']

  - job_name: node
    proxy_url: http://dev-k3s-2:8080/
    static_configs:
      - targets: ['dev-k3s-2.omni:9100']
```

**Comment by valyala** [comment]
@ThomasADavis , could you build `vmagent` from the commit 500e625e8c8896ba43eae0ab2c6e53bccdbf7a5a and verify whether it properly scrapes PushProx? The `vmagent` can be built from this commit with the following sequence of commands:

```
git clone https://github.com/VictoriaMetrics/VictoriaMetrics
cd VictoriaMetrics
git checkout 500e625e8c8896ba43eae0ab2c6e53bccdbf7a5a
make vmagent-prod
```

The `vmagent-prod` binary will be put in `bin` directory. See [these build instructions](https://victoriametrics.github.io/vmagent.html#how-to-build-from-sources) for more details.

**Comment by ThomasADavis** [comment]
Closer, different error.  using tcpdump to capture traffic between the proxy and vmagent/prometheus

```
16:09:15.025017 IP (tos 0x0, ttl 64, id 45459, offset 0, flags [DF], proto TCP (6), length 326)
    192.168.87.56.41644 > 192.168.87.56.webcache: Flags [P.], cksum 0x30fa (incorrect -> 0xc062), seq 269141510:269141784, ack 1720218806, win 114, options [nop,nop,TS val 748966872 ecr 748952103], length 274: HTTP, length: 274
	GET http://dev-k3s-2.omni:9100/metrics HTTP/1.1
	Host: dev-k3s-2.omni:9100
	User-Agent: Prometheus/2.20.0
	Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1
	Accept-Encoding: gzip
	X-Prometheus-Scrape-Timeout-Seconds: 10.000000
	
16:09:15.025049 IP (tos 0x0, ttl 64, id 21993, offset 0, flags [DF], proto TCP (6), length 52)
    192.168.87.56.webcache > 192.168.87.56.41644: Flags [.], cksum 0x2fe8 (incorrect -> 0xd054), seq 1, ack 274, win 272, options [nop,nop,TS val 748966872 ecr 748966872], length 0
16:09:15.025473 IP (tos 0x0, ttl 64, id 2254, offset 0, flags [DF], proto TCP (6), length 486)
    192.168.87.56.webcache > 192.168.87.56.41610: Flags [P.], cksum 0x319a (incorrect -> 0xec91), seq 3959703972:3959704406, ack 1935356754, win 272, options [nop,nop,TS val 748966872 ecr 748951872], length 434: HTTP, length: 434
	HTTP/1.1 200 OK
	Date: Sun, 04 Apr 2021 23:09:15 GMT
	Content-Length: 316
	Content-Type: text/plain; charset=utf-8
	
	GET http://dev-k3s-2.omni:9100/metrics HTTP/1.1
	Host: dev-k3s-2.omni:9100
	User-Agent: Prometheus/2.20.0
	Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1
	Accept-Encoding: gzip
	Id: aa281ac0-0e40-496e-ad11-3eebcad98802
	X-Prometheus-Scrape-Timeout-Seconds: 10.000000
```

and vmagent:

```
16:25:56.099945 IP (tos 0x0, ttl 64, id 59137, offset 0, flags [DF], proto TCP (6), length 223)
    192.168.87.56.51730 > 192.168.87.56.webcache: Flags [P.], cksum 0x3093 (incorrect -> 0x3e69), seq 223430172:223430343, ack 143484932, win 22, options [nop,nop,TS val 749967964 ecr 749952965], length 171: HTTP, length: 171
	GET http://dev-k3s-2.omni:9100/metrics HTTP/1.1
	User-Agent: vm_promscrape
	Host: dev-k3s-2:8080
	Accept: text/plain;version=0.0.4;q=1,*/*;q=0.1
	Accept-Encoding: gzip
	
16:25:56.100913 IP (tos 0x0, ttl 64, id 48896, offset 0, flags [DF], proto TCP (6), length 283)
    192.168.87.56.webcache > 192.168.87.56.51730: Flags [P.], cksum 0x30cf (incorrect -> 0x8608), seq 1:232, ack 171, win 23, options [nop,nop,TS val 749967965 ecr 749967964], length 231: HTTP, length: 231
	HTTP/1.1 500 Internal Server Error
	X-Prometheus-Scrape-Timeout: -9223372036.854776
	Date: Sun, 04 Apr 2021 23:25:56 GMT
	Content-Length: 46
	Content-Type: text/plain; charset=utf-8
	
	strconv.ParseFloat: parsing "": invalid syntax[!http]
```

Notice the different 'accept' line.  Not sure if that is actually the problem.
also notice the extra header for the prometheus timeout.

**Comment by ThomasADavis** [comment]
Ah, found it.

Needs to send the timeout header added.  That's this:

```
X-Prometheus-Scrape-Timeout-Seconds: 10.000000
```

from the client/main.go code:

```
func (c *Coordinator) doScrape(request *http.Request, client *http.Client) {
        logger := log.With(c.logger, "scrape_id", request.Header.Get("id"))
        timeout, err := util.GetHeaderTimeout(request.Header)
        if err != nil {
                c.handleErr(request, client, err)
                return
        }
```

from utils/proxy.go:

```
func GetHeaderTimeout(h http.Header) (time.Duration, error) {
	timeoutSeconds, err := strconv.ParseFloat(h.Get("X-Prometheus-Scrape-Timeout-Seconds"), 64)
	if err != nil {
		return time.Duration(0 * time.Second), err
	}

	return time.Duration(timeoutSeconds * 1e9), nil
}
```

**Comment by ThomasADavis** [comment]
and I can confirm, the missing X-Prometheus-Scrape-Timeout-Seconds is the final problem. 

I did this as a test:

```
[root@fn VictoriaMetrics]# git diff
diff --git a/lib/promscrape/client.go b/lib/promscrape/client.go
index 565e691..2d5314a 100644
--- a/lib/promscrape/client.go
+++ b/lib/promscrape/client.go
@@ -191,6 +191,7 @@ func (c *client) ReadData(dst []byte) ([]byte, error) {
        // See https://github.com/VictoriaMetrics/VictoriaMetrics/issues/608 for details.
        // Do not bloat the `Accept` header with OpenMetrics shit, since it looks like dead standard now.
        req.Header.Set("Accept", "text/plain;version=0.0.4;q=1,*/*;q=0.1")
+        req.Header.Set("X-Prometheus-Scrape-Timeout-Seconds", "10")
        if c.authHeader != "" {
                req.Header.Set("Authorization", c.authHeader)
        }
```

and now it works:

```
 [root@dev-k3s-2 ~]#  curl http://127.0.0.1:8429/targets
 job="node" (1/1 up) 
        state=up, endpoint=http://dev-k3s-2.omni:9100/metrics, labels={instance="dev-k3s-2.omni:9100",job="node"} , last_scrape=6.581s ago, scrape_duration=0.289s, error="" 
   job="prometheus" (1/1 up) 
        state=up, endpoint=http://localhost:9090/metrics, labels={instance="localhost:9090",job="prometheus"} , last_scrape=3.876s ago, scrape_duration=0.005s, error="" 
   
 [root@dev-k3s-2 ~]# 
```

**Comment by valyala** [comment]
@ThomasADavis , thanks for finding the root cause of the 500 error in PushProx! It should be fixed in the commit 6742839fd636b416f89a72a9e1f91107be3f6bb3 .

**Comment by ThomasADavis** [comment]
Looks good, waiting for the next release!
